sopel/db.py

```diff
@@ -180,9 +185,44 @@ def __init__(self, config):
     def connect(self):
         """Get a direct database connection.
 
-        :rtype: :class:`sqlalchemy.engine.Connection`
+        :return: a proxied DBAPI connection object; see
+                 :meth:`sqlalchemy.engine.Engine.raw_connection()`
+
+        .. important::
+
+           The :attr:`database backend <sopel.config.core_section.db_type>` in
+           use can change how the raw connection object behaves. Unless you're
+           writing (or updating) a plugin that needs to be compatible with
+           Sopel versions older than 7.0, or can get away with supporting only
+           one type of database, you probably want to use :meth:`session` and
+           build (or convert) your plugin to use the SQLAlchemy ORM.
+
+        """
+        if self.type != 'sqlite':
+            # log non-sqlite uses of raw connections for troubleshooting, since
+            # unless the developer had a good reason to use this instead of
+            # `session()`, it indicates the plugin was written before Sopel 7.0
+            # and might not work right when connected to non-sqlite DBs
+            LOGGER.info(
+                "Raw connection requested when db_type is not sqlite\n%s",
+                traceback.format_list(traceback.extract_stack()[:-1])[-1][:-1])
```

Comment:
I was about to say "use `stack_info`", but sadly, this feature has been added in Python 3.2 and is not available in Python 2.7 (yet another reason to complain about it).

So, I'd suggest to add a `# TODO: replace by 'stack_info=True' when Python 2.7 is dropped` above that line.

Reply:
I stole it from here: https://github.com/sopel-irc/sopel/blob/c5cffdb5443c5b8b3eda99a8c17fe1fdd3d3314d/sopel/tools/__init__.py#L350-L351

Using `stack_info=True` would give us the whole stack. We don't want the whole stack. We want just the most recent call, which is what this snippet yields.

Reply:
Well, it's hard to know in what context the call is made when you don't have the full context - that's what my experience tells me. But heh. I know you don't like traceback as much as I do. :p